File: common/src/main/java/cc/turtl/chiselmon/feature/pc/eggspy/EggDummy.java
package cc.turtl.chiselmon.feature.pc.eggspy;

import cc.turtl.chiselmon.ChiselmonConstants;
import cc.turtl.chiselmon.api.duck.DuckPreviewPokemon;
import com.cobblemon.mod.common.api.pokemon.feature.IntSpeciesFeature;
import com.cobblemon.mod.common.pokemon.Pokemon;
import com.cobblemon.mod.common.pokemon.RenderablePokemon;
import com.cobblemon.mod.common.pokemon.Species;
import net.minecraft.client.Minecraft;
import net.minecraft.core.RegistryAccess;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.TagParser;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;

import java.util.HashSet;
import java.util.Optional;
import java.util.Set;

/**
 * Representation of the pokemon a NeoDaycare egg will hatch into.
 *<p>
 * To most methods, will look like a regular pokemon object, only with
 * a forced aspect "EggDummy" for recognition by Chiselmon methods.
 */
public class EggDummy extends Pokemon {
    public static final String DUMMY_ASPECT = "EggDummy";
    public static final ResourceLocation EGG_SPECIES_ID = ResourceLocation.fromNamespaceAndPath("neodaycare", "egg_species");
    public static final String HATCH_PERCENTAGE_FEATURE = "hatch_percentage";

    private final Pokemon originalEgg;
    private int totalSteps;

    public EggDummy(Pokemon originalEgg) {
        this.originalEgg = originalEgg;
    }

    // Static factory method: The "New Parser"
    public static Optional<EggDummy> from(Pokemon egg) {
        String eggData = egg.getPersistentData().getString("Egg");
        if (eggData.isEmpty()) return Optional.empty();

        try {
            CompoundTag hatchlingNbt = TagParser.parseTag(eggData);
            EggDummy dummy = new EggDummy(egg);

            dummy.totalSteps = egg.getPersistentData().getInt("TotalSteps");

            if (Minecraft.getInstance().level == null) return Optional.empty();
            RegistryAccess registries = Minecraft.getInstance().level.registryAccess();

            // Pass the registries to the load method
            dummy.loadFromNBT(registries, hatchlingNbt);

            // Sync identity and hatch progress
            dummy.setUuid(egg.getUuid());
            Set<String> aspects = new HashSet<>(dummy.getForcedAspects());
            aspects.add(DUMMY_ASPECT);
            dummy.setForcedAspects(aspects);

            return Optional.of(dummy);
        } catch (Exception e) {
            ChiselmonConstants.LOGGER.error("Failed to parse hatchling for egg: {}", egg.getUuid(), e);
            return Optional.empty();
        }
    }

    public Pokemon getOriginalEgg() {
        return originalEgg;
    }

    /**
     * Pulled from the original Egg Pokemon object's 'features' list.
     * Returns 0 if no hatch percentage feature is found.
     */
    public int getHatchPercentage() {
        IntSpeciesFeature hatchFeature = this.originalEgg.getFeature(HATCH_PERCENTAGE_FEATURE);
        if (hatchFeature != null) {
            return hatchFeature.getValue();
        }
        return 0;
    }

    public int getTotalSteps() {
        return totalSteps;
    }

    public int getStepsRemaining() {
        return totalSteps - (totalSteps * getHatchPercentage() / 100);
    }

    public RenderablePokemon getOriginalRenderablePokemon() {
        return ((DuckPreviewPokemon) originalEgg).chiselmon$getRawRenderablePokemon();
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/feature/eggspy/EggDummy.kt
package cc.turtl.chiselmon.client.feature.eggspy

import cc.turtl.chiselmon.client.api.duck.DuckPreviewPokemon
import cc.turtl.chiselmon.core.ChiselmonConstants
import cc.turtl.chiselmon.mixin.accessor.AccessorPokemon
import com.cobblemon.mod.common.api.abilities.Ability
import com.cobblemon.mod.common.pokemon.Pokemon
import com.cobblemon.mod.common.pokemon.RenderablePokemon
import com.cobblemon.mod.common.util.DataKeys
import net.minecraft.client.Minecraft
import net.minecraft.nbt.NbtOps
import net.minecraft.nbt.TagParser
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.ItemStack

/**
 * Represents the pokemon a NeoDaycare egg will hatch into.
 *
 * To most methods, this looks like a regular Pokemon, with a forced
 * "EggDummy" aspect for recognition by Chiselmon methods.
 */
class EggDummy private constructor(
    private val cachedRenderableEgg: RenderablePokemon
) : Pokemon() {

    var totalSteps: Int = 0
        internal set

    var hatchPercentage: Int = 0
        internal set

    val stepsRemaining: Int
        get() = totalSteps - (totalSteps * hatchPercentage / 100)

    val originalRenderablePokemon: RenderablePokemon
        get() = cachedRenderableEgg

    private var cachedHatchlingRenderable: RenderablePokemon? = null

    /**
     * Cached [RenderablePokemon] for the hatchling. EggDummy fields are immutable
     * for its lifetime (EggCache rebuilds the dummy when hatchling data changes),
     * so the renderable can be reused every frame without wasting performance.
     */
    fun hatchlingRenderable(): RenderablePokemon =
        cachedHatchlingRenderable ?: RenderablePokemon(species, aspects, ItemStack.EMPTY)
            .also { cachedHatchlingRenderable = it }

    override fun attemptAbilityUpdate() {
        // Ability comes from hatchling data and should never be recalculated
    }

    companion object {
        const val DUMMY_ASPECT = "EggDummy"

        @JvmField
        val EGG_SPECIES_ID: ResourceLocation = ResourceLocation.fromNamespaceAndPath("neodaycare", "egg_species")
        const val HATCH_PERCENTAGE_FEATURE = "hatch_percentage"

        fun from(egg: Pokemon): EggDummy? {
            val eggData = egg.persistentData.getString("Egg")
            if (eggData.isEmpty()) return null

            return try {
                val hatchlingNbt = TagParser.parseTag(eggData)
                val registries = Minecraft.getInstance().level?.registryAccess() ?: return null
                val ops = registries.createSerializationContext(NbtOps.INSTANCE)
                val renderableEgg = (egg as DuckPreviewPokemon).`chiselmon$getRawRenderablePokemon`()

                val decoded = CODEC.decode(ops, hatchlingNbt)
                    .result()
                    .orElse(null)
                    ?.getFirst() ?: return null

                val formId = hatchlingNbt.getString(DataKeys.POKEMON_FORM_ID)

                EggDummy(renderableEgg).also { dummy ->
                    dummy.copyFrom(decoded)

                    // Explicitly restore the actual form. My original solution lost form data like typing
                    // and only preserved visual form aspects
                    val resolvedForm = dummy.species.forms
                        .firstOrNull { it.formOnlyShowdownId() == formId }

                    if (resolvedForm != null) {
                        dummy.form = resolvedForm
                    }

                    // The codec decode chain corrupts the ability so it needs to be reapplied.
                    Ability.CODEC.decode(NbtOps.INSTANCE, hatchlingNbt.getCompound(DataKeys.POKEMON_ABILITY))
                        .result()
                        .orElse(null)
                        ?.getFirst()
                        ?.let {
                            @Suppress("CAST_NEVER_SUCCEEDS")
                            (dummy as AccessorPokemon).`chiselmon$setAbility`(it)
                        }

                    // Lock client mode so updateAspects just uses forcedAspects (preserves the form aspects set above)
                    @Suppress("CAST_NEVER_SUCCEEDS")
                    (dummy as AccessorPokemon).`chiselmon$setIsClient`(true)

                    val capturedAspects =
                        decoded.aspects + (resolvedForm?.aspects ?: emptySet())

                    dummy.forcedAspects = capturedAspects + DUMMY_ASPECT
                    dummy.uuid = egg.uuid

                    // Prefix the display label with (EGG) so render-site mixins that just
                    // read getDisplayName() / getickname() don't have to know about egg logic.
                    val baseName = dummy.nickname?.copy() ?: dummy.species.translatedName.copy()
                    dummy.nickname = Component.literal("(EGG) ").append(baseName)
                }
            } catch (e: Exception) {
                ChiselmonConstants.LOGGER.error("Failed to parse hatchling for egg: {}", egg.uuid, e)
                null
            }
        }
    }
}